File: Seori/Weekly Contest 404/3201. Find the Maximum Length of Valid Subsequence I.py
class Solution:
    def maximumLength(self, nums: List[int]) -> int:

        # [1] nums의 각 원소를 2로 나눈 나머지를 리스트로 만든다. Make a list of the remainder of each element of nums divided by 2.
        nums = list(map(lambda x: x % 2, nums))

        # [2] 짝수, 홀수, alternate를 세는 리스트를 만든다. Make a list that counts even, odd, and alternating numbers.
        count = [0, 0, 1] # even, odd, alternate

        # [3] nums의 원소를 하나씩 확인하여 짝수/홀수 및 alternate인지를 확인한다. Check if the element of nums is even/odd and alternate one by one.
        now = nums[0]
        count[now % 2] += 1
        for i in range(1, len(nums)):
            next = nums[i]
            # count odd or even
            count[next % 2] += 1
            # count alternate
            if (now + next) % 2:
                count[2] += 1
            now = next
        
        # [4] 짝수, 홀수, alternate 중 가장 큰 값을 반환한다. Return the largest value among even, odd, and alternate.
        return max(count)

File: Gombang/Weekly Contest 404/3201_Find_the_Maximum_Length_of_Valid_Subsequence_I.cs
public class Solution
{
    public int MaximumLength(int[] nums)
    {
        int even = 0, odd = 0;
        int toggleState = nums[0] % 2;
        int toggleCount = 0;

        // 반복문을 돌리면서 값이 짝수면 even값을 ++, 홀수이면 odd값을 ++.
        // 만약 짝수와 홀수가 번갈아 나오는 상황에 대해 체크하기 위해서 toggleState로 
        // 해당 상황을 체크하며, toggleCount값을 증가시킴.
        foreach (int num in nums)
        {
            if (num % 2 == 0)
                even++;
            else
                odd++;

            if (num % 2 == toggleState)
            {
                toggleCount++;
                toggleState = 1 - toggleState;
            }
        }

        return Math.Max(toggleCount, Math.Max(odd, even));
    }
}